lightning/src/routing/scoring.rs

```diff
@@ -678,25 +795,103 @@ impl<L: Deref<Target = u64>, T: Time, U: Deref<Target = T>> DirectedChannelLiqui
+			// historical liquidity bucket (min, max) combinations (where the min is less than the
+			// max) and, for each, calculate the probability of success given our payment amount,
+			// then total the weighted average probability of success.
+			//
+			// We use a sliding scale to decide which value will be used for a given bucket - the
+			// first bucket receives zero (ie the lower-bound of the bucket), the last bucket
+			// receives nine (ie the upper-bound of the bucket). This avoids failing to assign
+			// penalties to channels at the edges.
+			//
+			// If we used the bottom edge of buckets, we'd end up never assigning any penalty at
+			// all to such a channel when sending less than ~0.19% of the channel's capacity (e.g.
+			// ~200k sats for a 1 BTC channel!).
+			//
+			// If we used the middle of each bucket we'd never assign any penalty at all when
+			// sending less than 1/16th of a channel's capacity, or 1/8th if we used the top of the
+			// bucket.
+			let mut total_valid_points_tracked = 0;
+			let decays = self.now.duration_since(*self.last_updated).as_secs()
+				.checked_div(params.historical_no_updates_half_life.as_secs())
+				.map_or(u32::max_value(), |decays| cmp::min(decays, u32::max_value() as u64) as u32);
+			// Rather than actually decaying the individual buckets, which would lose precision, we
+			// simply track whether any buckets have data which wouldn't be decayed to zero, and if
+			// there are none, treat it as if we had no data.
+			let mut is_fully_decayed = true;
+			for (min_idx, min_bucket) in self.min_liquidity_offset_history.buckets.iter().enumerate() {
+				if min_bucket.checked_shr(decays).unwrap_or(0) > 0 { is_fully_decayed = false; }
+				for max_bucket in self.max_liquidity_offset_history.buckets.iter().take(8 - min_idx) {
+					total_valid_points_tracked += (*min_bucket as u64) * (*max_bucket as u64);
+					if max_bucket.checked_shr(decays).unwrap_or(0) > 0 { is_fully_decayed = false; }
+				}
+			}
+			if total_valid_points_tracked.checked_shr(decays).unwrap_or(0) < 31*31 || is_fully_decayed {
+				// If we don't have any valid points (or, once decayed, we have less than a full
+				// point), redo the non-historical calculation with no liquidity bounds tracked and
+				// the historical penalty multipliers.
+				let max_capacity = self.capacity_msat.saturating_sub(amount_msat).saturating_add(1);
+				let negative_log10_times_2048 =
+					approx::negative_log10_times_2048(max_capacity, self.capacity_msat.saturating_add(1));
+				res = res.saturating_add(Self::combined_penalty_msat(amount_msat, negative_log10_times_2048,
+					params.historical_liquidity_penalty_multiplier_msat,
+					params.historical_liquidity_penalty_amount_multiplier_msat));
+				return res;
+			}
+
+			let payment_amt_64th_bucket = amount_msat * 64 / self.capacity_msat;
+			debug_assert!(payment_amt_64th_bucket <= 64);
+			if payment_amt_64th_bucket > 64 { return res; }
```

Comment:
Why the combination of assert `<=` and return if `>`?

Reply:
Yes, I just wasn't clear on what we can expect the value to be? Will it always be `<= 64`, or won't it? What would be the condition that would make us to return here in the prod case?